File: packages/core/src/auth/xai-oauth.ts
import { createServer } from "node:http";
import { XaiErrorCode, XaiOAuthError } from "./errors";
import type { XaiOAuthCredentials } from "../types";

const DEFAULT_BASE_URL = "https://api.x.ai/v1";
const ISSUER = "https://auth.x.ai";
const DISCOVERY_URL = `${ISSUER}/.well-known/openid-configuration`;
const CLIENT_ID = process.env.PI_XAI_OAUTH_CLIENT_ID || "b1a00492-073a-47ea-816f-4c329264a828";
const SCOPE =
  process.env.PI_XAI_OAUTH_SCOPE ||
  "openid profile email offline_access grok-cli:access api:access";
const CALLBACK_HOST = process.env.PI_XAI_OAUTH_CALLBACK_HOST || "127.0.0.1";
const CALLBACK_PORT = Number.parseInt(process.env.PI_XAI_OAUTH_CALLBACK_PORT || "56121", 10);
const CALLBACK_PATH = "/callback";
const REFRESH_SKEW_MS = 120_000;

export interface XaiDiscovery {
  authorization_endpoint: string;
  token_endpoint: string;
}

export interface OAuthLoginCallbacks {
  onAuth(info: { url: string; instructions: string }): void;
}

export function getBaseUrl(): string {
  return (
    process.env.PI_XAI_BASE_URL ||
    process.env.XAI_BASE_URL ||
    DEFAULT_BASE_URL
  ).replace(/\/+$/, "");
}

function base64Url(buffer: ArrayBuffer | Uint8Array): string {
  const bytes = buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer);
  let binary = "";
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

async function generatePKCE(): Promise<{ verifier: string; challenge: string }> {
  const verifier = base64Url(crypto.getRandomValues(new Uint8Array(32)));
  const hash = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(verifier));
  return { verifier, challenge: base64Url(hash) };
}

function validateEndpoint(value: string, field: string): string {
  let url: URL;
  try {
    url = new URL(value);
  } catch {
    throw new XaiOAuthError(
      `xAI OAuth discovery returned invalid ${field}: ${value}`,
      XaiErrorCode.DISCOVERY_INVALID_ORIGIN,
    );
  }
  if (url.protocol !== "https:") {
    throw new XaiOAuthError(`xAI OAuth ${field} must use HTTPS: ${value}`, XaiErrorCode.DISCOVERY_INVALID_ORIGIN);
  }
  const host = url.hostname.toLowerCase();
  if (host !== "x.ai" && host !== "auth.x.ai" && host !== "accounts.x.ai" && !host.endsWith(".x.ai")) {
    throw new XaiOAuthError(`Refusing non-xAI OAuth ${field}: ${value}`, XaiErrorCode.DISCOVERY_INVALID_ORIGIN);
  }
  return url.toString();
}

export async function discover(): Promise<XaiDiscovery> {
  let response: Response;
  try {
    response = await fetch(DISCOVERY_URL, {
      headers: { Accept: "application/json" },
      signal: AbortSignal.timeout(15_000),
    });
  } catch (cause) {
    throw new XaiOAuthError(
      `xAI OIDC discovery failed: ${cause instanceof Error ? cause.message : String(cause)}`,
      XaiErrorCode.DISCOVERY_FAILED,
    );
  }
  if (!response.ok) {
    throw new XaiOAuthError(`xAI OIDC discovery returned ${response.status}`, XaiErrorCode.DISCOVERY_FAILED);
  }
  const payload = (await response.json()) as Record<string, unknown>;
  return {
    authorization_endpoint: validateEndpoint(String(payload.authorization_endpoint ?? ""), "authorization_endpoint"),
    token_endpoint: validateEndpoint(String(payload.token_endpoint ?? ""), "token_endpoint"),
  };
}

interface CallbackResult {
  code?: string;
  state?: string;
  error?: string;
  errorDescription?: string;
}

function startCallbackServer() {
  let settle: ((value: CallbackResult) => void) | undefined;
  const callbackPromise = new Promise<CallbackResult>((resolve) => {
    settle = resolve;
  });

  const server = createServer((req, res) => {
    try {
      const origin = req.headers.origin;
      if (origin === "https://accounts.x.ai" || origin === "https://auth.x.ai") {
        res.setHeader("Access-Control-Allow-Origin", origin);
        res.setHeader("Access-Control-Allow-Methods", "GET, OPTIONS");
        res.setHeader("Access-Control-Allow-Headers", "Content-Type");
        res.setHeader("Access-Control-Allow-Private-Network", "true");
        res.setHeader("Vary", "Origin");
      }
      if (req.method === "OPTIONS") {
        res.statusCode = 204;
        res.end();
        return;
      }
      const url = new URL(req.url ?? "/", `http://${CALLBACK_HOST}`);
      if (url.pathname !== CALLBACK_PATH) {
        res.statusCode = 404;
        res.end("Not found");
        return;
      }
      const result: CallbackResult = {
        code: url.searchParams.get("code") ?? undefined,
        state: url.searchParams.get("state") ?? undefined,
        error: url.searchParams.get("error") ?? undefined,
        errorDescription: url.searchParams.get("error_description") ?? undefined,
      };
      res.statusCode = result.error ? 400 : 200;
      res.setHeader("Content-Type", "text/html; charset=utf-8");
      const html = result.error
        ? "<html><body><h1>xAI authorization failed.</h1>You can close this tab.</body></html>"
        : "<html><body><h1>xAI authorization received.</h1>You can close this tab.</body></html>";
      res.end(html);
      settle?.(result);
    } catch {
      res.statusCode = 500;
      res.end("Internal error");
    }
  });

  const listen = (port: number) =>
    new Promise<number>((resolve, reject) => {
      server.once("error", reject);
      server.listen(port, CALLBACK_HOST, () => {
        server.removeListener("error", reject);
        const addr = server.address();
        resolve(typeof addr === "object" && addr ? addr.port : port);
      });
    });

  return (async () => {
    let actualPort: number;
    try {
      actualPort = await listen(CALLBACK_PORT);
    } catch {
      actualPort = await listen(0);
    }
    const redirectUri = `http://${CALLBACK_HOST}:${actualPort}${CALLBACK_PATH}`;
    return {
      server,
      redirectUri,
      waitForCallback: (timeoutMs: number) =>
        Promise.race([
          callbackPromise,
          new Promise<CallbackResult>((resolve) =>
            setTimeout(
              () => resolve({ error: "timeout", errorDescription: "Timed out waiting for xAI OAuth callback." }),
              timeoutMs,
            ),
          ),
        ]),
    };
  })();
}

async function exchangeCode(
  tokenEndpoint: string,
  code: string,
  redirectUri: string,
  verifier: string,
): Promise<XaiOAuthCredentials> {
  const response = await fetch(tokenEndpoint, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded", Accept: "application/json" },
    body: new URLSearchParams({
      grant_type: "authorization_code",
      client_id: CLIENT_ID,
      code,
      redirect_uri: redirectUri,
      code_verifier: verifier,
    }),
  });
  if (!response.ok) {
    throw new XaiOAuthError(
      `xAI token exchange failed: ${response.status} ${await response.text()}`,
      XaiErrorCode.TOKEN_EXCHANGE_FAILED,
    );
  }
  const payload = (await response.json()) as Record<string, unknown>;
  const access = String(payload.access_token ?? "");
  const refresh = String(payload.refresh_token ?? "");
  if (!access) {
    throw new XaiOAuthError("xAI token exchange did not return access_token.", XaiErrorCode.TOKEN_EXCHANGE_INVALID);
  }
  if (!refresh) {
    throw new XaiOAuthError("xAI token exchange did not return refresh_token.", XaiErrorCode.TOKEN_EXCHANGE_INVALID);
  }
  const expiresIn = typeof payload.expires_in === "number" ? payload.expires_in : Number(payload.expires_in ?? 3600);
  return {
    access,
    refresh,
    expires: Date.now() + expiresIn * 1000 - REFRESH_SKEW_MS,
    tokenEndpoint,
    discovery: { authorization_endpoint: "", token_endpoint: tokenEndpoint },
    idToken: String(payload.id_token ?? ""),
    tokenType: String(payload.token_type ?? "Bearer"),
    baseUrl: getBaseUrl(),
  };
}

export async function login(callbacks: OAuthLoginCallbacks): Promise<XaiOAuthCredentials> {
  const discovery = await discover();
  const { verifier, challenge } = await generatePKCE();
  const state = base64Url(crypto.getRandomValues(new Uint8Array(16)));
  const nonce = base64Url(crypto.getRandomValues(new Uint8Array(16)));
  const callback = await startCallbackServer();

  try {
    const authUrl = new URL(discovery.authorization_endpoint);
    authUrl.searchParams.set("response_type", "code");
    authUrl.searchParams.set("client_id", CLIENT_ID);
    authUrl.searchParams.set("redirect_uri", callback.redirectUri);
    authUrl.searchParams.set("scope", SCOPE);
    authUrl.searchParams.set("code_challenge", challenge);
    authUrl.searchParams.set("code_challenge_method", "S256");
    authUrl.searchParams.set("state", state);
    authUrl.searchParams.set("nonce", nonce);
    authUrl.searchParams.set("plan", "generic");
    authUrl.searchParams.set("referrer", "pi-agent");

    callbacks.onAuth({
      url: authUrl.toString(),
      instructions: `Authorize xAI, then return to pi-agent. Callback listener: ${callback.redirectUri}`,
    });

    const result = await callback.waitForCallback(180_000);
    if (result.error) {
      throw new XaiOAuthError(result.errorDescription ?? result.error, XaiErrorCode.AUTHORIZATION_FAILED);
    }
    if (result.state !== state) {
      throw new XaiOAuthError("xAI OAuth state mismatch — possible CSRF.", XaiErrorCode.STATE_MISMATCH);
    }
    if (!result.code) {
      throw new XaiOAuthError("xAI OAuth callback did not include an authorization code.", XaiErrorCode.CODE_MISSING);
    }
    const credentials = await exchangeCode(discovery.token_endpoint, result.code, callback.redirectUri, verifier);
    credentials.discovery = discovery;
    return credentials;
  } finally {
    callback.server.close();
  }
}

export async function refresh(credentials: XaiOAuthCredentials): Promise<XaiOAuthCredentials> {
  const tokenEndpoint =
    credentials.tokenEndpoint || credentials.discovery?.token_endpoint || (await discover()).token_endpoint;
  validateEndpoint(tokenEndpoint, "token_endpoint");

  if (!credentials.refresh) {
    throw new XaiOAuthError("Missing refresh_token. Re-login required.", XaiErrorCode.REFRESH_MISSING, true);
  }

  const response = await fetch(tokenEndpoint, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded", Accept: "application/json" },
    body: new URLSearchParams({
      grant_type: "refresh_token",
      client_id: CLIENT_ID,
      refresh_token: credentials.refresh,
    }),
  });

  if (!response.ok) {
    const isFatal = response.status === 400 || response.status === 401 || response.status === 403;
    throw new XaiOAuthError(
      `xAI token refresh failed: ${response.status} ${await response.text()}`,
      XaiErrorCode.REFRESH_FAILED,
      isFatal,
    );
  }

  const payload = (await response.json()) as Record<string, unknown>;
  const access = String(payload.access_token ?? "");
  if (!access) {
    throw new XaiOAuthError("xAI token refresh did not return access_token.", XaiErrorCode.REFRESH_FAILED, true);
  }
  const refresh_new = String(payload.refresh_token ?? credentials.refresh);
  const expiresIn = typeof payload.expires_in === "number" ? payload.expires_in : Number(payload.expires_in ?? 3600);

  return {
    ...credentials,
    access,
    refresh: refresh_new,
    expires: Date.now() + expiresIn * 1000 - REFRESH_SKEW_MS,
    tokenEndpoint,
    idToken: String(payload.id_token ?? credentials.idToken ?? ""),
    tokenType: String(payload.token_type ?? credentials.tokenType ?? "Bearer"),
    baseUrl: getBaseUrl(),
  };
}

File: internal/modules/core/config/xai_oauth.go
package config

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// Signing in to xAI with a SuperGrok subscription, rather than with an API
// key.
//
// The distinction matters commercially: an API key is pay-as-you-go, while
// the OAuth path bills against a subscription the user already has. Offering
// only the key means a SuperGrok subscriber pays twice, which is why loop
// offers both and why this exists.
//
// It is the standard authorization-code flow with PKCE, and the parts that
// are not standard are the parts that keep it safe on a developer's machine:
//
//   - The redirect is a loopback listener on 127.0.0.1, so the code never
//     leaves the machine. There is no client secret — a public client cannot
//     keep one — and PKCE is what stands in for it.
//   - Every endpoint comes from OIDC DISCOVERY and is then checked to be
//     HTTPS and to belong to x.ai. Discovery means a URL the user's DNS
//     resolves, and following one of those to a token endpoint without
//     checking it is how an access token ends up somewhere else.
//   - `state` is compared on the way back. Without it the callback accepts a
//     code from anyone who can reach the listener.

const (
	xaiIssuer    = "https://auth.x.ai"
	xaiDiscovery = xaiIssuer + "/.well-known/openid-configuration"
	// The public client id loop registered. A public client's id is not a
	// secret — PKCE is what makes the flow safe without one.
	xaiClientID     = "b1a00492-073a-47ea-816f-4c329264a828"
	xaiScope        = "openid profile email offline_access grok-cli:access api:access"
	xaiCallbackHost = "127.0.0.1"
	// The port xAI has registered for the loopback redirect. A different one
	// is only used if this is taken, and then the redirect must still be
	// accepted by the authorization server.
	xaiCallbackPort = 56121
	xaiCallbackPath = "/callback"
	// refreshSkew renews a token before it actually expires, so a request
	// cannot lose a race with the clock.
	refreshSkew = 2 * time.Minute
)

// XaiCreds is what the flow yields and what is stored.
type XaiCreds struct {
	Access        string `json:"access"`
	Refresh       string `json:"refresh"`
	Expires       int64  `json:"expires"`
	TokenEndpoint string `json:"tokenEndpoint"`
	TokenType     string `json:"tokenType,omitempty"`
}

type xaiEndpoints struct {
	Authorization string `json:"authorization_endpoint"`
	Token         string `json:"token_endpoint"`
}

// XaiLogin runs the browser flow and stores the credentials.
//
// present is called with the URL to open and a line explaining what is about
// to happen — the caller owns the terminal, and this package must not print.
func XaiLogin(ctx context.Context, present func(url, instructions string)) error {
	endpoints, err := xaiDiscover(ctx)
	if err != nil {
		return err
	}
	verifier, challenge, err := pkce()
	if err != nil {
		return err
	}
	state, err := randomToken(16)
	if err != nil {
		return err
	}
	nonce, err := randomToken(16)
	if err != nil {
		return err
	}

	cb, err := startCallbackServer()
	if err != nil {
		return err
	}
	defer cb.close()

	auth, err := url.Parse(endpoints.Authorization)
	if err != nil {
		return fmt.Errorf("xai: bad authorization endpoint: %w", err)
	}
	q := auth.Query()
	q.Set("response_type", "code")
	q.Set("client_id", xaiClientID)
	q.Set("redirect_uri", cb.redirectURI)
	q.Set("scope", xaiScope)
	q.Set("code_challenge", challenge)
	q.Set("code_challenge_method", "S256")
	q.Set("state", state)
	q.Set("nonce", nonce)
	q.Set("plan", "generic")
	q.Set("referrer", "pi-agent")
	auth.RawQuery = q.Encode()

	present(auth.String(), "Authorize xAI in the browser, then come back here. Listening on "+cb.redirectURI)

	result, err := cb.wait(ctx, 3*time.Minute)
	if err != nil {
		return err
	}
	if result.state != state {
		return fmt.Errorf("xai: state mismatch — the callback did not come from the sign-in that was started")
	}
	creds, err := xaiExchange(ctx, endpoints.Token, result.code, cb.redirectURI, verifier)
	if err != nil {
		return err
	}
	return saveXaiCreds(creds)
}

// XaiAccessToken returns a usable bearer token, refreshing it when it has
// expired and writing the new one back.
func XaiAccessToken(ctx context.Context) (string, error) {
	creds, ok := loadXaiCreds()
	if !ok {
		return "", fmt.Errorf("not signed in to xAI")
	}
	if time.Now().UnixMilli() < creds.Expires {
		return creds.Access, nil
	}
	fresh, err := xaiRefresh(ctx, creds)
	if err != nil {
		return "", err
	}
	if err := saveXaiCreds(fresh); err != nil {
		return "", err
	}
	return fresh.Access, nil
}

// XaiSignedIn reports whether a subscription login is stored.
func XaiSignedIn() bool {
	_, ok := loadXaiCreds()
	return ok
}

func xaiDiscover(ctx context.Context) (xaiEndpoints, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, xaiDiscovery, nil)
	if err != nil {
		return xaiEndpoints{}, err
	}
	req.Header.Set("Accept", "application/json")
	resp, err := (&http.Client{Timeout: 15 * time.Second}).Do(req)
	if err != nil {
		return xaiEndpoints{}, fmt.Errorf("xai: discovery failed: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return xaiEndpoints{}, fmt.Errorf("xai: discovery returned %d", resp.StatusCode)
	}
	var e xaiEndpoints
	if err := json.NewDecoder(resp.Body).Decode(&e); err != nil {
		return xaiEndpoints{}, fmt.Errorf("xai: discovery is not readable: %w", err)
	}
	if err := validateXaiEndpoint(e.Authorization, "authorization_endpoint"); err != nil {
		return xaiEndpoints{}, err
	}
	if err := validateXaiEndpoint(e.Token, "token_endpoint"); err != nil {
		return xaiEndpoints{}, err
	}
	return e, nil
}

// validateXaiEndpoint refuses anything that is not HTTPS and not x.ai.
//
// Discovery is a document fetched over the network. Following whatever it
// names — to a token endpoint, with an authorization code in hand — is how a
// compromised resolver walks off with an account.
func validateXaiEndpoint(raw, field string) error {
	u, err := url.Parse(raw)
	if err != nil || u.Host == "" {
		return fmt.Errorf("xai: discovery returned an invalid %s: %q", field, raw)
	}
	if u.Scheme != "https" {
		return fmt.Errorf("xai: %s must be https: %q", field, raw)
	}
	host := strings.ToLower(u.Hostname())
	if host != "x.ai" && !strings.HasSuffix(host, ".x.ai") {
		return fmt.Errorf("xai: refusing a non-x.ai %s: %q", field, raw)
	}
	return nil
}

type callbackResult struct {
	code, state string
	err         string
}

type callbackServer struct {
	srv         *http.Server
	ln          net.Listener
	redirectURI string
	results     chan callbackResult
}

func startCallbackServer() (*callbackServer, error) {
	ln, err := net.Listen("tcp", fmt.Sprintf("%s:%d", xaiCallbackHost, xaiCallbackPort))
	if err != nil {
		// The registered port is taken — another sign-in, or another tool.
		// Any port still works provided the server accepts the redirect.
		ln, err = net.Listen("tcp", xaiCallbackHost+":0")
		if err != nil {
			return nil, fmt.Errorf("xai: cannot listen for the callback: %w", err)
		}
	}
	cb := &callbackServer{
		ln:      ln,
		results: make(chan callbackResult, 1),
		redirectURI: fmt.Sprintf("http://%s:%d%s",
			xaiCallbackHost, ln.Addr().(*net.TCPAddr).Port, xaiCallbackPath),
	}
	mux := http.NewServeMux()
	mux.HandleFunc(xaiCallbackPath, func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()
		res := callbackResult{code: q.Get("code"), state: q.Get("state")}
		if e := q.Get("error"); e != "" {
			res.err = e
			if d := q.Get("error_description"); d != "" {
				res.err = d
			}
		}
		w.Header().Set("Content-Type", "text/html; charset=utf-8")
		if res.err != "" {
			w.WriteHeader(http.StatusBadRequest)
			w.Write([]byte("<h1>xAI authorization failed.</h1><p>You can close this tab.</p>"))
		} else {
			w.Write([]byte("<h1>xAI authorization received.</h1><p>You can close this tab and return to pi-agent.</p>"))
		}
		select {
		case cb.results <- res:
		default:
		}
	})
	cb.srv = &http.Server{Handler: mux}
	go cb.srv.Serve(ln)
	return cb, nil
}

func (c *callbackServer) wait(ctx context.Context, timeout time.Duration) (callbackResult, error) {
	select {
	case res := <-c.results:
		if res.err != "" {
			return res, fmt.Errorf("xai: authorization failed: %s", res.err)
		}
		if res.code == "" {
			return res, fmt.Errorf("xai: the callback carried no authorization code")
		}
		return res, nil
	case <-ctx.Done():
		return callbackResult{}, ctx.Err()
	case <-time.After(timeout):
		return callbackResult{}, fmt.Errorf("xai: timed out waiting for the browser")
	}
}

func (c *callbackServer) close() {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()
	_ = c.srv.Shutdown(ctx)
}

func xaiExchange(ctx context.Context, tokenEndpoint, code, redirectURI, verifier string) (XaiCreds, error) {
	form := url.Values{
		"grant_type":    {"authorization_code"},
		"client_id":     {xaiClientID},
		"code":          {code},
		"redirect_uri":  {redirectURI},
		"code_verifier": {verifier},
	}
	payload, err := postForm(ctx, tokenEndpoint, form)
	if err != nil {
		return XaiCreds{}, fmt.Errorf("xai: token exchange failed: %w", err)
	}
	access, _ := payload["access_token"].(string)
	refresh, _ := payload["refresh_token"].(string)
	if access == "" {
		return XaiCreds{}, fmt.Errorf("xai: token exchange returned no access token")
	}
	if refresh == "" {
		// Without one the session dies at the first expiry and the user is
		// sent back through the browser with no explanation.
		return XaiCreds{}, fmt.Errorf("xai: token exchange returned no refresh token")
	}
	return XaiCreds{
		Access:        access,
		Refresh:       refresh,
		Expires:       expiryFrom(payload),
		TokenEndpoint: tokenEndpoint,
		TokenType:     stringOr(payload["token_type"], "Bearer"),
	}, nil
}

func xaiRefresh(ctx context.Context, creds XaiCreds) (XaiCreds, error) {
	endpoint := creds.TokenEndpoint
	if endpoint == "" {
		e, err := xaiDiscover(ctx)
		if err != nil {
			return XaiCreds{}, err
		}
		endpoint = e.Token
	}
	// Re-checked even though it was checked when stored: the file is on disk
	// and editable, and this is the request that carries the refresh token.
	if err := validateXaiEndpoint(endpoint, "token_endpoint"); err != nil {
		return XaiCreds{}, err
	}
	if creds.Refresh == "" {
		return XaiCreds{}, fmt.Errorf("xai: no refresh token — sign in again")
	}
	payload, err := postForm(ctx, endpoint, url.Values{
		"grant_type":    {"refresh_token"},
		"client_id":     {xaiClientID},
		"refresh_token": {creds.Refresh},
	})
	if err != nil {
		return XaiCreds{}, fmt.Errorf("xai: token refresh failed: %w", err)
	}
	access, _ := payload["access_token"].(string)
	if access == "" {
		return XaiCreds{}, fmt.Errorf("xai: token refresh returned no access token")
	}
	out := creds
	out.Access = access
	out.Expires = expiryFrom(payload)
	out.TokenEndpoint = endpoint
	// A rotated refresh token replaces the old one; a server that does not
	// rotate simply omits it, and the existing one stays valid.
	if r, ok := payload["refresh_token"].(string); ok && r != "" {
		out.Refresh = r
	}
	return out, nil
}

func postForm(ctx context.Context, endpoint string, form url.Values) (map[string]any, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	resp, err := (&http.Client{Timeout: 30 * time.Second}).Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var payload map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return nil, fmt.Errorf("status %d, and the body is not readable", resp.StatusCode)
	}
	if resp.StatusCode != http.StatusOK {
		// The provider's own message, which is the only one that says what is
		// actually wrong — an expired grant, a revoked client.
		if msg, ok := payload["error_description"].(string); ok && msg != "" {
			return nil, fmt.Errorf("status %d: %s", resp.StatusCode, msg)
		}
		return nil, fmt.Errorf("status %d", resp.StatusCode)
	}
	return payload, nil
}

func expiryFrom(payload map[string]any) int64 {
	seconds := 3600.0
	if v, ok := payload["expires_in"].(float64); ok && v > 0 {
		seconds = v
	}
	return time.Now().Add(time.Duration(seconds)*time.Second - refreshSkew).UnixMilli()
}

func stringOr(v any, fallback string) string {
	if s, ok := v.(string); ok && s != "" {
		return s
	}
	return fallback
}

func pkce() (verifier, challenge string, err error) {
	verifier, err = randomToken(32)
	if err != nil {
		return "", "", err
	}
	sum := sha256.Sum256([]byte(verifier))
	return verifier, base64.RawURLEncoding.EncodeToString(sum[:]), nil
}

func randomToken(n int) (string, error) {
	buf := make([]byte, n)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(buf), nil
}

// loadXaiCreds reads the OAuth entry from the shared auth file.
func loadXaiCreds() (XaiCreds, bool) {
	doc, err := readAuthDoc()
	if err != nil {
		return XaiCreds{}, false
	}
	providers, _ := doc["providers"].(map[string]any)
	entry, _ := providers["xai"].(map[string]any)
	if entry == nil || entry["mode"] != "oauth" {
		return XaiCreds{}, false
	}
	raw, err := json.Marshal(entry["xai"])
	if err != nil {
		return XaiCreds{}, false
	}
	var creds XaiCreds
	if err := json.Unmarshal(raw, &creds); err != nil || creds.Access == "" {
		return XaiCreds{}, false
	}
	return creds, true
}

func saveXaiCreds(creds XaiCreds) error {
	doc, err := readAuthDoc()
	if err != nil {
		doc = map[string]any{}
	}
	providers, _ := doc["providers"].(map[string]any)
	if providers == nil {
		providers = map[string]any{}
	}
	providers["xai"] = map[string]any{
		"mode": "oauth", "provider": "xai", "xai": creds,
	}
	doc["providers"] = providers
	if _, ok := doc["active"]; !ok {
		doc["active"] = "xai"
	}
	return writeAuthDoc(doc)
}

// LogoutXai removes the subscription credentials.
func LogoutXai() error {
	doc, err := readAuthDoc()
	if err != nil {
		return fmt.Errorf("no stored credentials")
	}
	providers, _ := doc["providers"].(map[string]any)
	if providers == nil {
		return fmt.Errorf("no stored credentials")
	}
	if _, ok := providers["xai"]; !ok {
		return fmt.Errorf("not signed in to xAI")
	}
	delete(providers, "xai")
	doc["providers"] = providers
	return writeAuthDoc(doc)
}
